server/broadcast.go

```diff
@@ -0,0 +1,214 @@
+	}
+	go func() {
+		// storage the orchestrator prefers
+		if ios := sess.OrchestratorOS; ios != nil {
+			// XXX handle case when orch expects direct upload
+			uri, err := ios.SaveData(name, seg.Data)
+			if err != nil {
+				glog.Error("Error saving segment to OS ", err)
+				if monitor.Enabled {
+					monitor.LogSegmentUploadFailed(nonce, seg.SeqNo, err.Error())
+				}
+				return
+			}
+			seg.Name = uri // hijack seg.Name to convey the uploaded URI
+		}
+
+		// send segment to the orchestrator
+		glog.V(common.DEBUG).Infof("Submitting segment %d", seg.SeqNo)
+
+		res, err := SubmitSegment(sess, seg, nonce)
+		if err != nil {
+			if shouldStopStream(err) {
+				glog.Warningf("Stopping current stream due to: %v", err)
+				rtmpStrm.Close()
+			}
+			if shouldStopSession(err) {
+				cxn.needOrch <- struct{}{}
+			}
+			return
+		}
+
+		// download transcoded segments from the transcoder
+		gotErr := false // only send one error msg per segment list
+		errFunc := func(subType, url string, err error) {
+			glog.Errorf("%v error with segment %v: %v (URL: %v)", subType, seg.SeqNo, err, url)
+			if monitor.Enabled && !gotErr {
+				monitor.LogSegmentTranscodeFailed(subType, nonce, seg.SeqNo, err)
+				gotErr = true
+			}
+		}
+
+		segHashes := make([][]byte, len(res.Segments))
+		n := len(res.Segments)
+		SegHashLock := &sync.Mutex{}
+		cond := sync.NewCond(SegHashLock)
+
+		dlFunc := func(url string, i int) {
+			defer func() {
+				cond.L.Lock()
+				n--
+				if n == 0 {
+					cond.Signal()
+				}
+				cond.L.Unlock()
+			}()
+
+			if bos := sess.BroadcasterOS; bos != nil && !drivers.IsOwnExternal(url) {
+				data, err := drivers.GetSegmentData(url)
+				if err != nil {
+					errFunc("Download", url, err)
+					return
+				}
+				name := fmt.Sprintf("%s/%d.ts", sess.Profiles[i].Name, seg.SeqNo)
+				newUrl, err := bos.SaveData(name, data)
+				if err != nil {
+					errFunc("SaveData", url, err)
+					return
+				}
+				url = newUrl
+
+				hash := crypto.Keccak256(data)
+				SegHashLock.Lock()
+				segHashes[i] = hash
+				SegHashLock.Unlock()
+			}
+
+			if monitor.Enabled {
+				monitor.LogTranscodedSegmentAppeared(nonce, seg.SeqNo, sess.Profiles[i].Name)
+			}
+			err = cpl.InsertHLSSegment(&sess.Profiles[i], seg.SeqNo, url, seg.Duration)
+			if err != nil {
+				errFunc("Playlist", url, err)
+				return
+			}
+		}
+
+		for i, v := range res.Segments {
+			go dlFunc(v.Url, i)
+		}
+
+		cond.L.Lock()
+		for n != 0 {
+			cond.Wait()
+		}
+		cond.L.Unlock()
+
+		// if !eth.VerifySig(transcoderAddress, crypto.Keccak256(segHashes...), res.Sig) { // need transcoder address here
+		// 	glog.Error("Sig check failed for segment ", seg.SeqNo)
+		// 	return
+		// }
+
+		glog.V(common.DEBUG).Info("Successfully validated segment ", seg.SeqNo)
+	}()
+}
+
+func generateSessionErrors() *regexp.Regexp {
+	// Given a list [err1, err2, err3] generates a regexp `(err1)|(err2)|(err3)`
+	errStrings := []string{"dial tcp", "unexpected EOF", "OrchestratorBusy"}
```

Comment:
I'm curious to learn:

- How did you compile this list?
- How do we know it's exhaustive?

Reply:
Manually tested the following conditions:
* Orchestrator down at initial connection / segment submission time yields an error with the substring `dial tcp`
* Orchestrator down mid-transcode yields an error with the substring `unexpected EOF`
* Orchestrator with a busy transcode loop yields an error with `OrchestratorBusy`

This list is by no means exhaustive, it's just what I've been able to capture in practice. If you can think of more scenarios that'd merit an immediate failover, or see additional cases during testing, this list can certainly be expanded.

Reply:
Perhaps it's interesting to test the case of O taking a very long time to respond to the HTTP request?

Reply:
The error on timeouts will look approximately like this:

``` rpc.go:459] Unable to submit segment 12 Post https://127.0.0.1:8936/segment: net/http: request canceled (Client.Timeout exceeded while awaiting headers)```

From our [earlier discussion](https://github.com/livepeer/go-livepeer/issues/636) it seemed like we settled on using a counter for timeouts. We can probably match on `Client.Timeout` . The current implementation doesn't incorporate a counter yet though; it only handles hard failovers.

Reply:
Agreed that timeouts fall under the issue you linked to. Do I understand correctly that in the current implementation we give full grace to timeouts? i.e. just keep sending segments regardless of how many timeouts occur?